node/src/chain/fallback.rs

```diff
@@ -0,0 +1,180 @@
+        let mut prev_block = ledger::Block::default();
+        acc.db.read().await.view(|v| {
+            prev_block = Ledger::fetch_block_by_height(&v, prev_block_height)?
+                .ok_or_else(|| anyhow::anyhow!("could not fetch block"))?;
+
+            Ok(())
+        })?;
+
+        info!("Verify block header/certificate data");
+
+        // Validate Header/Certificate of the new block upon previous block and
+        // provisioners.
+
+        // In an edge case, this may fail on performing fallback between two
+        // epochs.
+        let provisioners_list = acc.provisioners_list.read().await;
+        acceptor::verify_block_header(
+            self.acc.db.clone(),
+            prev_block.header(),
+            provisioners_list.clone(),
+            &PublicKey::default(),
+            blk.header(),
+        )
+        .await
+    }
+
+    async fn execute_fallback(&self, blk: &Block) -> Result<()> {
+        let acc = self.acc;
+        let curr_height = acc.get_curr_height().await;
+        let curr_iteration = acc.get_curr_iteration().await;
+
+        info!("Revert VM to last finalized state");
+        let state_hash_after_revert = acc.vm.read().await.revert()?;
+
+        info!(
+            "Revert completed, finalized_state_hash:{}",
+            hex::ToHex::encode_hex::<String>(&state_hash_after_revert)
+        );
+
+        // Delete any ephemeral block until we reach the last finalized block,
+        // the VM was reverted to.
+        info!("Delete all most recent ephemeral blocks");
+
+        // Thew blockchain tip (most recent block) after reverting to last
+        // finalized state.
+        let mut new_mrb = Block::default();
+
+        acc.db.read().await.update(|t| {
+            let mut height = curr_height;
+            loop {
+                if height == 0 {
+                    break;
+                }
+
+                let chain_blk = Ledger::fetch_block_by_height(t, height)?
+                    .ok_or_else(|| anyhow::anyhow!("could not fetch block"))?;
+
+                let iteration = chain_blk.header.iteration;
+                if chain_blk.header.state_hash == state_hash_after_revert {
+                    info!(
+                        "state_hash found at height: {}, iter: {}",
+                        height, iteration
+                    );
+
+                    new_mrb = chain_blk;
+                    break;
+                }
+
+                if iteration == 1 {
+                    /// A sanity check to ensure we never delete a finalized
+                    /// block
+                    warn!("deleting a block from first iteration");
+                }
+
+                Ledger::delete_block(t, &chain_blk)?;
+
+                // Attempt to resubmit transactions back to mempool.
+                // An error here is not considered critical.
+                for tx in blk.txs().iter().skip(1) {
+                    // TODO: Filter out Distribute tx explicitly
```

Comment:
I don't think we still have `distribute` txs

Reply:
Good catch!